uploader.go

```diff
@@ -184,22 +192,47 @@ func (u UploadOptions) contextWithTimeout(ctx context.Context) (context.Context,
 	return context.WithTimeout(ctx, u.BatchWindow)
 }
 
+func addToPendingBatch(u UploadOptions, inputData InputData) Batch {
+	blob := inputData.GetBlob()
+	objectType := inputData.GetObjectType()
+	b, ok := pendingBatches[objectType]
+
+	// start a new batch
+	if !ok {
+		firstBlobTime := time.Now()
+		b = Batch{
+			firstBlobTime: firstBlobTime,
+			size:          uint64(len(blob)),
+			key:           inputData.GetKey(firstBlobTime),
+			objectType:    objectType,
+		}
+
+		b.buffer = <-u.bufferPool
+		pendingBatches[objectType] = b
+	} else {
+		b.size += uint64(len(blob))
+	}
+
+	b.buffer = append(b.buffer, blob)
+	return b
+}
+
+func (b Batch) close() {
+	delete(pendingBatches, b.objectType)
+}
+
 // getBatch collects a batch together and returns it and additional metadata. Respects the various
 // batch parameters, including BatchSizeBytes, BatchWindow, and BatchMaxBlobs.
 func (u UploadOptions) getBatch(runCtx context.Context) *Batch {
-	blob, ok := <-u.Input
+	inputData, ok := <-u.Input
 	if !ok {
 		return nil
 	}
 
-	b := Batch{
-		firstBlobTime: time.Now(),
-		size:          uint64(len(blob)),
-	}
-	b.buffer = <-u.bufferPool
-	b.buffer = append(b.buffer, blob)
+	b := addToPendingBatch(u, inputData)
```

Comment:
This copies the `Batch` struct; any `append` done to the `buffer` won't be reflected in the `pendingBatches` map.